File: TypeScript/01-binarySearch.ts
abstract class Searcher {
    abstract search(inputList: (number | string)[], searchElement: number | string): number | null;
}

class BinarySearch extends Searcher {
    search(inputList: (number | string)[], searchElement: number | string): number | null {
        const iL = inputList;
        const sE = searchElement;
        if (iL.length > 1) {
            console.log(iL.length, iL);
            const middleElementIdx = Math.floor(iL.length / 2);
            const middleElement = iL[middleElementIdx]

            if (middleElement == sE) {
                return middleElementIdx;
            } else if (middleElement < sE) {
                const result = this.search(iL.slice(middleElementIdx),sE)
                return result !== null ? middleElementIdx + result : null
            } else {
                return this.search(iL.slice(0, middleElementIdx),sE)
            }
        }
        return null
    }
}

class SearchFactory {
    private _searchHistory: Searcher[] = [];

    createSearcher(searcherType: string): Searcher {
        switch (searcherType) {
            case 'binary':
                const newSearcher = new BinarySearch();
                this._searchHistory.push(newSearcher);
                return newSearcher;
            default:
                throw new Error("No proper Searcher selected")
        }
    }
}

const alist = [3, 8, 12, 15, 22, 31, 50];
const aEl = 50;

const searchFactory = new SearchFactory();
const searcher = searchFactory.createSearcher("binary");
console.log(searcher.search(alist, aEl));
console.log("nice");

File: Python/01-binarySearch.py
from abc import abstractmethod
from typing import Any, List


class Searcher:
    @abstractmethod
    def search(self,inputList:List[Any],searchElement:int|float|str)->int|None:
        pass

class BinarySearch(Searcher):
    def search(self, inputList: List[Any],searchElement:int|float|str) -> int:
        iL = inputList
        sE = searchElement
        if len(iL)>1:
            print(len(iL),iL)
            middleElementIdx = len(iL)//2
            middleElement=iL[middleElementIdx]
            if middleElement == sE:
                return middleElementIdx
            elif middleElement < sE:
                return middleElementIdx + self.search(iL[middleElementIdx:],sE)
            return self.search(iL[:middleElementIdx],sE)
        return -len(iL)-1

class SearchFactory:
    def __init__(self,searcher_history:List[Searcher]=[]):
        self._searcher_history = searcher_history
    def create_searcher(self,searcher:str):
        match searcher:
            case "binary": 
                new_searcher = BinarySearch()
                self._searcher_history.append(new_searcher)
                return new_searcher
            case _:
                raise ValueError("No proper Searcher selected.")

alist = [3, 8, 12, 15, 22, 31, 50]
aEl = 9

search_factory = SearchFactory()
searcher = search_factory.create_searcher("binary")
print(searcher.search(alist,aEl))
print("nice")
